### ISSUE TITLE
Inference with pre-trained ensembled nnUNet 

### ISSUE THREAD

**Issue by PaWeRe** [DIAGNijmegen/picai_baseline#46]
Hi,

I would like to reproduce the results in the pi-cai forum ([here](https://grand-challenge.org/forums/forum/pi-cai-607/topic/baseline-ai-models-for-3d-cspca-detectiondiagnosis-in-bpmri-updates-and-fixes-715/)) on the Public Training and Development Dataset, by using the pre-trained nnUNet models provided ([picai_nnunet_gc_algorithm](https://github.com/DIAGNijmegen/picai_nnunet_gc_algorithm), [picai_nnunet_semi_supervised_gc_algorithm](https://github.com/DIAGNijmegen/picai_nnunet_semi_supervised_gc_algorithm)). 

Following the instructions on the github page ([nnunet_baseline.md](https://github.com/DIAGNijmegen/picai_baseline/blob/main/nnunet_baseline.md)), could you help me figure out how to consolidate the pre-trained folds, to then run inference (as specified in "or ensemble the models for the test set")? It seems that in contrast to the nnDetection documentation, there is no consolidate script provided in the picai_baseline repo?

Maybe this might also be interesting to other people, hence worth updating the documentation :)

Thank you for your time!

Best,

Patrick

**Comment by joeranbosma** [comment]
Hi @PaWeRe,

Thanks for pointing out this is unclear. The models that are ensembled are the “model_best” checkpoints for each of the five folds. This is the configuration from the inference code ([proces.py](https://github.com/DIAGNijmegen/picai_nnunet_semi_supervised_gc_algorithm/blob/eb4add5770805f019597aac5d27629bbf5279d3c/process.py#L168)) and the [inference command](https://github.com/DIAGNijmegen/picai_baseline/blob/main/nnunet_baseline.md#nnu-net---inference), but I don't think it's mentioned explicitly. 

I hope this helps you set up the models as intended. If you have a specific change you think would make it clearer, a PR is always appreciated!

Kind regards,
Joeran

**Comment by joeranbosma** [comment]
If you have additional concerns, feel free to let me know. Closing this as complete for now.

**Comment by bibhabasumohapatra** [comment]
@joeranbosma thanks for the repository. it supports the folder structures of folders t2,adc,hbv transverse.
and creates intermediates of 0000,0001, and 0002.
Actually I wanted to infer the model on other public datasets. using the process.py preprocess section I preprocessed the data into required spacing, crop and others.

Now I have a folder with all images, id_A_0000,id_0001,id_B_0002. Which also created in line:
https://github.com/DIAGNijmegen/picai_nnunet_semi_supervised_gc_algorithm/blob/6129f7e0fe3fc3994064672aa2c7f629584004a3/process.py#L146

I want to use your model after that to see how it performs, but I am unable change or understand the code after that.
Can you let me know a small format where preprocessed inputs 0000,0001, and 0002 are taken as input to the nnUnet model to get nii.gz output.

I am asking for a lot but, even little guidance will be helpful for me to change the code

**Comment by joeranbosma** [comment]
Dear @bibhabasumohapatra,

I hope I understand your question correctly, do let me know if I misunderstood. 

Do you want to get out the predicted detection map (as opposed to e.g. the predicted softmax volume)?
After inference, the model prediction can be found in a file like this: `images/cspca-detection-map/cspca_detection_map.mha`.

For local inference with a similar algorithm, I've used the script below. In this case, the preprocessing is part of the process.py, that's the same here right? (depending on how you want to use this baseline). Is this helpful to answer your question?

Alternatively, or as a debugging step for a couple of cases, you can also use the version on Grand Challenge which takes care of all steps: https://grand-challenge.org/algorithms/pi-cai-baseline-nnu-net-semi-supervised/

Note: for this baseline algorithm, the clinical variables are not used, so copying those can also be left out.


```python
import argparse
import shutil
from pathlib import Path

import SimpleITK as sitk
from picai_prep.data_utils import atomic_file_copy, atomic_image_write
from process import csPCaAlgorithm as Algorithm

if __name__ == "__main__":
    # Command line options
    parser = argparse.ArgumentParser(description="Batch inference options", formatter_class=argparse.ArgumentDefaultsHelpFormatter)
    parser.add_argument("--input_dir", type=Path, default=Path("/dataset"),
                        help="Path to the input directory")
    parser.add_argument("--output_dir", type=Path, default=Path("/results"),
                        help="Path to the output directory")
    parser.add_argument("--algorithm_input", type=Path, default=Path("/input"),
                        help="Path to the algorithm input directory")
    parser.add_argument("--algorithm_output", type=Path, default=Path("/output"),
                        help="Path to the algorithm output directory")
    parser.add_argument("--subject_list", type=Path, default=None,
                        help="Path to the subject list stored in a text file (one subject_id per line)")
    args = parser.parse_args()

    # Set up paths
    input_dir = args.input_dir
    images_dir = input_dir / "images"
    output_dir = args.output_dir
    algorithm_input = args.algorithm_input
    algorithm_output = args.algorithm_output

    workdirs = [
        algorithm_input,
        algorithm_output,
    ]

    print(f"Input dir: {input_dir}")
    print(f"Images dir: {images_dir}")
    print(f"Output dir: {output_dir}")

    if args.subject_list is not None:
        with open(args.subject_list) as f:
            subject_list = [line.strip() for line in f.readlines()]
        print(f"Using subject list from {args.subject_list}")
    else:
        input_scans = sorted(list(images_dir.glob("*/*_t2w.mha")))
        subject_list = [path.name.split("_t2w.mha")[0] for path in input_scans]
        print(f"Found {len(input_scans)} cases")

    print("Starting inference...")
    for subject_id in subject_list:
        print("="*100)
        print(f"Starting {subject_id}...")
        patient_id = subject_id.split("_")[0]
        path_detection_map_dst = output_dir / f"cspca-detection-map/{subject_id}.nii.gz"
        path_case_pred_dst = output_dir / f"cspca-case-level-likelihood/{subject_id}.json"
        if path_detection_map_dst.exists() and path_case_pred_dst.exists():
            print(f"Skipping {subject_id} because output already exists")
            continue

        try:
            # clean up
            for path in workdirs:
                shutil.rmtree(path, ignore_errors=True)
                path.mkdir(exist_ok=True, parents=True)

            # prepare input dir on host computer
            for num, (folder, modality) in enumerate([
                ("images/transverse-t2-prostate-mri", "t2w"),
                ("images/transverse-adc-prostate-mri", "adc"),
                ("images/transverse-hbv-prostate-mri", "hbv"),
            ]):
                src = images_dir / patient_id / f"{subject_id}_{modality}.mha"
                dst = algorithm_input / folder / f"{num}_{subject_id}_{modality}.mha"
                print(f"Copying {src} to {dst}...")
                atomic_file_copy(src, dst, mkdir=True)

            # prepare input dir on host computer
            src = input_dir / "clinical_information_v2" / f"{subject_id}.json"
            dst = algorithm_input / "clinical-information-prostate-mri.json"
            atomic_file_copy(src, dst, mkdir=True)

            # run algorithm
            Algorithm().predict()

            # copy detection map out
            src = algorithm_output / "images/cspca-detection-map/cspca_detection_map.mha"
            pred = sitk.ReadImage(str(src))
            atomic_image_write(pred, path_detection_map_dst, mkdir=True)

            # copy case-level prediction out
            src = algorithm_output / "cspca-case-level-likelihood.json"
            atomic_file_copy(src, path_case_pred_dst, mkdir=True)
        except Exception as e:
            print(f"Error for {subject_id}: {e}")
            continue
```


This script expects the data under /dataset to be ordered in the same manner as the PI-CAI dataset.

**Comment by bibhabasumohapatra** [comment]
thanks for a quick reply!!  I will try this once for sure. Actually Sorry, I saw the message now. (maybe closed issue, so didn't get GitHub notification)

**Comment by bibhabasumohapatra** [comment]
@joeranbosma thanks for the script I still cannot figure out the folder's name (esp: algorithm_input and algorithm_output), though I see that the script you shared above is more helpful than process.py.

What I have is a folder with preprocessed outputs: 
```
/data/ImageTr/Patient1_0000.nii.gz, 
/data/ImageTr/Patient1_0001.nii.gz,
/data/ImageTr/Patient1_0002.nii.gz,
/data/ImageTr/Patient2_0000.nii.gz,  
/data/ImageTr/Patient2_0001.nii.gz,
/data/ImageTr/Patient2_0002.nii.gz,   
```
and this checkpoint folder from :
https://github.com/DIAGNijmegen/picai_nnunet_semi_supervised_gc_algorithm/tree/master/results/nnUNet/3d_fullres/Task2203_picai_baseline/nnUNetTrainerV2_Loss_FL_and_CE_checkpoints__nnUNetPlansv2.1

is it possible for me to infer using baseline weights?
